### PR TITLE
Extended generateStructuredMesh

### PR DESCRIPTION
**Pull request by chleh**
* some arguments that are mutually exclusive are checked
* --mx now works together with --nx etc.
* the existing subdivision class now checks if (infinitely many) subdivisions can sum up to the total length.

### CODE REVIEW

File: BaseLib/Error.h [hunk 1]
```diff
@@ -13,25 +13,51 @@
...
-    {\
-    ERR("%s", BaseLib::format(fmt, ##__VA_ARGS__).data());\
-    std::abort();\
-    }
+namespace BaseLib
+{
+namespace detail
+{
+template <typename Msg>
+[[noreturn]] bool error_impl(Msg&& msg)
```
Comment (endJunction) [inline BaseLib/Error.h]: :+1:

File: BaseLib/Subdivision.cpp [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+            dL0,
+            multiplier,
+            L);
+    }
+}
+
+GradualSubdivisionFixedNum::GradualSubdivisionFixedNum(
+    const double L, const std::size_t num_subdivisions, const double multiplier)
+    : _length{L}, _num_subdivisions{num_subdivisions}, _multiplier{multiplier}
+{
```
Comment (endJunction) [inline BaseLib/Subdivision.cpp]: can this be `= default`?

**Comment by chleh** [comment]
Appveyor error: `ERROR: Missing prebuilt package for 'VTK/7.1.0@bilke/stable'`
🆘 Can somebody help me?

**Comment by endJunction** [comment]
Regarding appveyor; sometimes the conan servers do not respond, so the package is not found. But seems to be working now.

File: Applications/Utils/SimpleMeshCreation/generateStructuredMesh.cpp [hunk 3]
```diff
@@ -179,17 +180,26 @@ int main (int argc, char* argv[])
     vec_div.reserve(dim);
     for (unsigned i=0; i<dim; i++)
     {
-        if (vec_ndivArg[i]->isSet())
-        {
+        if (vec_multiArg[i]->isSet()) {
+            if (vec_ndivArg[i]->isSet()) {
+                // number of partitions in direction is specified
+                vec_d0Arg[i]->isSet() &&
```
Comment (TomFischer) [inline Applications/Utils/SimpleMeshCreation/generateStructuredMesh.cpp]: Is `vec_d0Arg[i]->isSet() && OGS_FATAL(...)`  another notation for
```
if (vec_d0Arg[i]->isSet())
    OGS_FATAL(...)
```
?

**Comment by TomFischer** [comment]
One reason the checks are not successful is a missing semicolon in `Applications/Utils/ModelPreparation/PartitionMesh/PartitionMesh.cpp +137`.

File: Applications/Utils/SimpleMeshCreation/generateStructuredMesh.cpp (hunk 3 above)
Comment (chleh) [inline Applications/Utils/SimpleMeshCreation/generateStructuredMesh.cpp]: Yes.

**Comment by chleh** [comment]
@TomFischer Thanks for the partmesh hint.

File: BaseLib/Subdivision.cpp (hunk 2 above)
Comment (chleh) [inline BaseLib/Subdivision.cpp]: I guess not since this is a multi-argument ctor.

**Pull request merged**